arangod/Agency/Store.cpp

```diff
@@ -614,51 +607,64 @@ bool Store::applies(arangodb::velocypack::Slice const& transaction) {
             });
 
   bool success = true;
+
+  auto node = _node;
+  TRI_ASSERT(node != nullptr);
   for (const auto& i : idx) {
     Slice value = i.second;
 
     if (value.isObject() && value.hasKey("op")) {
       std::string_view const op = value.get("op").stringView();
 
       if (op == "delete" || op == "replace" || op == "erase") {
-        if (!_node->has(abskeys.at(i.first))) {
+        if (!node->has(abskeys.at(i.first))) {
           continue;
         }
       }
-      auto uri = Store::normalize(abskeys.at(i.first));
-      auto ret = _node->hasAsWritableNode(abskeys.at(i.first)).applyOp(value);
-      success &= ret.ok();
+
+      auto ret = node->applyOp(abskeys.at(i.first), value);
+      if (ret.fail()) {
+        success = false;
+        break;
+      }
+      node = std::move(ret).get();
+      if (node == nullptr) {  // if root node is deleted, keep an empty node
+        node = Node::create();
+      }
       // check for triggers here
       callTriggers(abskeys.at(i.first), op, value);
     } else {
-      success &= _node->hasAsWritableNode(abskeys.at(i.first)).applies(value);
+      node = node->applies(abskeys.at(i.first), value);
```

Comment:
I know this has been like this before, but triggers will not be checked in this branch of the if. That is, one can avoid triggers by using the shortcut without `"op"`, I think we should fix this right here.